modules/os2forms_nemid/os2forms_nemid.module

```diff
@@ -71,12 +83,78 @@ function os2forms_nemid_webform_third_party_settings_form_alter(&$form, FormStat
     '#default_value' => !(empty($settings)) ? $settings['nemlogin_auto_redirect'] : FALSE,
     '#description' => t('Redirection will happen right after user has is accessing the form, if user is already authenticated via NemID, redirection will not happen.'),
   ];
+
+  $nemloginProtectionSettings = $webform->getThirdPartySetting('os2forms', 'os2forms_nemid_address_protection');
+
+  // OS2Forms NemID.
+  $form['third_party_settings']['os2forms']['os2forms_nemid_address_protection'] = [
+    '#type' => 'details',
+    '#title' => t('OS2Forms address protection settings'),
+    '#open' => TRUE,
+  ];
+
+  // Nemlogin auto redirect.
+  $form['third_party_settings']['os2forms']['os2forms_nemid_address_protection']['nemlogin_hide_form'] = [
+    '#type' => 'select',
+    '#options' => [
+      FormsHelper::WEBFORM_NEM_LOGIN_ADDRESS_PROTECTION_DEFAULT_BEHAVIOUR => t('No'),
+      FormsHelper::WEBFORM_NEM_LOGIN_ADDRESS_PROTECTION_DISPLAY_ERROR => t('Yes'),
+    ],
+    '#title' => t('Hide form if user is under protection'),
+    '#default_value' => !(empty($nemloginProtectionSettings)) ? $nemloginProtectionSettings['nemlogin_hide_form'] : FormsHelper::WEBFORM_NEM_LOGIN_ADDRESS_PROTECTION_DEFAULT_BEHAVIOUR,
+    '#description' => t('Hides elements and displays error if nemlogin reveals that citizen is under address protection and an address element is found on the webform'),
+  ];
+
+  // Nemlogin address protection.
+  $form['third_party_settings']['os2forms']['os2forms_nemid_address_protection']['nemlogin_hide_message'] = [
+    '#title' => t('Hide message'),
+    '#type' => 'textarea',
+    '#default_value' => !(empty($nemloginProtectionSettings)) ? $nemloginProtectionSettings['nemlogin_hide_message'] : '',
+    '#description' => t('Message shown to user when visiting form'),
+    '#states' => [
+      'visible' => [
+        [':input[name="third_party_settings[os2forms][os2forms_nemid_address_protection][nemlogin_hide_form]"]' => ['value' => FormsHelper::WEBFORM_NEM_LOGIN_ADDRESS_PROTECTION_DISPLAY_ERROR]],
+      ],
+      'required' => [
+        [':input[name="third_party_settings[os2forms][os2forms_nemid_address_protection][nemlogin_hide_form]"]' => ['value' => FormsHelper::WEBFORM_NEM_LOGIN_ADDRESS_PROTECTION_DISPLAY_ERROR]],
+      ],
+    ],
+  ];
 }
 
 /**
  * Implements hook_webform_submission_form_alter().
  */
 function os2forms_nemid_webform_submission_form_alter(array &$form, FormStateInterface $form_state, $form_id) {
+
+  // Handle address protection.
+  if ($tempValue = $form_state->getTemporaryValue(FormsHelper::TEMPORARY_KEY)) {
+    if (FALSE === ($tempValue['access'] ?? TRUE)) {
+      // Flattening the elements makes it much easier to access nested elements.
+      $elements = &WebformFormHelper::flattenElements($form['elements']);
+
+      \Drupal::messenger()->addError(t('Access to form denied'));
```

Comment:
This message is a little harsh. Something along the lines of “This form cannot be shown because you have address protection” is nicer.

Consider only showing the message if no custom message is set:

```php
        $message = $tempValue['message'] ?? t('Access to form denied');
        $form['os2forms_nemlogin_message'] = [
          '#theme' => 'status_messages',
          '#message_list' => [
            'error' => [$tempValue['message']],
          ],
        ];
```

Reply:
Reworded and use only the default message if no custom message is configured.